File: Cpp solutions/1046 last stone weight.cpp
// https://leetcode.com/problems/last-stone-weight
// TS:O(n^2logn) SC:O(1)
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        while (stones.size() > 1){
            sort(stones.begin(), stones.end());
            int smashResult = stones[stones.size() - 1] - stones[stones.size() - 2];
            if (smashResult != 0){
                stones.insert(stones.begin(), smashResult);
            }
            stones.erase(stones.begin() + (stones.size() - 2));
            stones.erase(stones.begin() + (stones.size() - 1));
        }
        if (stones.size() == 0){
            return 0;
        }
        else{
            return stones[0];
        }
    }
};

File: Python solutions/1046 last stone weight.py
# https://leetcode.com/problems/last-stone-weight
# TC: o(n*log(n)) SC: o(n)
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        max_heap = []
        for n in stones:
            heappush(max_heap, -n)
        while len(max_heap) > 1:
            first = heappop(max_heap)
            second = heappop(max_heap)
            if first != second:
                heappush(max_heap, first - second)
        return -max_heap[0] if max_heap else 0

